document supported C++ standards and compilers (#67)
BlueBrain/hpc-coding-conventions: cpp/Tooling.md

## C++ Compiler

### C++ Standard

The minimal C++ standard officially supported at BBP is C++14, to be reviewed annually based on compiler support.
An open source project may support a standard as low as C++11 if this is a requirement for its community.

> BBP internal deployment toolchain supports C++17. Therefore software dedicated to BB5 or BlueBrain workstations can pick this standard.

Reviewer: ... and do not require the Intel compilers for optimal performance ...

Author: You are right. I will remove the last sentence until we have an intel compiler deployed on BB5 that fully support C++17. Thanks

### Compiler Versions Supporting C++14

* GCC 5.4
* CLang 7
* Intel 18.0.3

### Compilation flags

> Here is the list of recommended compilation flags you may use during the development phase in order to spot as much errors as possible. Use CMake function `bob_begin_cxx_flags` provided by this CMake project to set the compilation flags accordingly.

Reviewer 2: ```suggestion
Here is the list of recommended compilation flags you may use during the development phase in order to spot as much errors as possible. For convenience, you may use the CMake function `bob_begin_cxx_flags` provided by this CMake project to set the compilation flags accordingly.
```

## C++ Compiler

### C++ Standard

> The C++ standard officially supported at BBP is C++14.

Reviewer 2: ```suggestion
The minimal C++ standard officially supported at BBP is C++14, to be reviewed annually based on compiler support.
```

> An source project can support C++11 if this is a requirement for its community.

Reviewer 3: typo?

Author: yes thank you. Do not hesitate to provide _suggestions_ that can be committed right-away.

> An source project can support C++11 if this is a requirement for its community.

Reviewer 2: ```suggestion
An source project may support a standard as low as C++11 if this is a requirement for its community.
```

BBP internal deployment toolchain supports C++17. Therefore software dedicated to BB5 or BlueBrain workstations can pick this standard.

> ### Minimum Supported Compilers

Reviewer 2: This sounds a little dangerous. Does this mean that all code should be tested with GCC5.4 to ensure support for that? As in, will all your compilation flags used in Robert work with all these compilers (and newer ones)?

I Iike
```suggestion
### Compiler Versions Supporting C++14
```

Author: IMO we should support GCC 5.4 for:
* open-source projects
* projects deployed on external super computers because as @pramodk dictated, we cannot expect to have the same toolchain than BB5.

Author: > As in, will all your compilation flags used in Robert work with all these compilers (and newer ones)?

I should clarify that the compilation flags mentioned below are the ones recommended for development, not deployment.

Reviewer 4: > because as @pramodk dictated 

thank you @tristan0x! 😊 

Matthias - it’s same as Python version. We are not testing/deploying with Python 3.6 version all the time but as developers we should make sure that code can be built with compiler supporting C++14.
